### PR TITLE
MF-838 - Migration to Go Module

### LINKED ISSUE #838: Migration to Go Module
**FEATURE REQUEST**

Switch to Go module makes the code more compatible with Go environment and standard.

### PR DESCRIPTION
**Pull request by 1995parham**
### What does this do?
It adds `go.mod` and `go.sum` into the project. Also, we can remove the Go Dep files.

### Which issue(s) does this PR fix/relate to?
Resolves #838 

### List any changes that modify/break current functionality
There are no breaking changes, but if we remove go dep file, we break the Go Dep compatibility.

### Notes
We can create specific `go.mod` for each independent component.

### CODE REVIEW

**Comment by drasko** [comment]
@1995parham Mainflux project includes vendoring (we keep `vendor` folder in our GitHub). Please make sure that Go mod will include vendoring as well.

Take care that you remove `Gopkg.lock` and `Gopkg.toml` files (previously generated by `dep` tool) when you do module generation, so you can start with fresh go mod files.

**Comment by 1995parham** [comment]
@drasko I will update the vendoring folder, and I will remove the dep files. Thanks for mentioning these.

**Comment by nmarcetic** [comment]
@1995parham Any progress here?  We are very excited about switching, so if we can help somehow, let us know. Thanks!

**Comment by 1995parham** [comment]
@nmarcetic I think it is ready and I have updated the vendor folder. If all contributors prepared for it, we can merge it. 😄

**Comment by anovakovic01** [comment]
@1995parham Please update your branch, sign your commits and remove WIP label when this PR is ready.

**Comment by 1995parham** [comment]
@1995parham I will do it right now

**Comment by drasko** [comment]
Is this PR ready? Can WIP be removed?

Also, please update the branch.

**Comment by 1995parham** [comment]
@drasko It is ready, I will update it.

File: go.mod
Comment (anovakovic01) [inline go.mod]: I recommend that you change this one to version `0.4.1`

Comment (1995parham) [reply go.mod]: @anovakovic01 Done

**Review by anovakovic01 — DISMISSED** [review]
LGTM!

**Review by drasko — DISMISSED** [review]
LGTM

**Review by dborovcanin — DISMISSED** [review]

**Comment by drasko** [comment]
@1995parham looks like PR got enught approvals. Please resolve the conflicts and update branch so we can merge it.

**Comment by 1995parham** [comment]
@drasko, I have updated my branch and resolve the conflicts.

File: scripts/ci.sh
Comment (drasko) [inline scripts/ci.sh]: We officially switched to Go1.13 (we needed this because of new error package).

I do not think that we need this Go1.11 ENV any more.

**Review by drasko — CHANGES_REQUESTED** [review]

File: scripts/ci.sh
Comment (1995parham) [reply scripts/ci.sh]: Without this package didn't install globally. To remove this, I think we need to install them with `go install` instead of `go get`.

Comment (drasko) [reply scripts/ci.sh]: Which Go version do you have on your machine? Can you switch to Go 1.13 and try?

Comment (1995parham) [reply scripts/ci.sh]: Yes, I will check it and let you know.

Comment (1995parham) [reply scripts/ci.sh]: I think here we want to install some packages in `GOPATH`, so we need to disable modules and install them. Another way is to change the current directory and then execute `go get` so it will install into `GOPATH`.

**Comment by nmarcetic** [comment]
@1995parham Can you resolve conflicts please ?

**Comment by 1995parham** [comment]
@nmarcetic I have resolved the conflicts, but there is an issue with MQTT library on the tests.

**Comment by dborovcanin** [comment]
> @nmarcetic I have resolved the conflicts, but there is an issue with MQTT library on the tests.

@1995parham Can you, please, update Paho lib version to `1.2.0`?

**Review by dborovcanin — DISMISSED** [review]
@1995parham Please rebase, so that we can merge this one. Great job!

File: scripts/ci.sh
Comment (drasko) [reply scripts/ci.sh]: @1995parham I would prefer either adding a comment that explains that, or actually doing that - because it is not self-documenting.

Adding a comment should be enough for now.

**Review by drasko — CHANGES_REQUESTED** [review]

**Comment by drasko** [comment]
@1995parham getting there. Only one comment left from my side, and I will approve.

File: scripts/ci.sh
Comment (drasko) [reply scripts/ci.sh]: We have an internal code-style convention that comments start with capital letter. This is neatpicking, but please correct the comment.

**Review by drasko — CHANGES_REQUESTED** [review]

**Review by drasko — APPROVED** [review]
LGTM

**Review by dborovcanin — DISMISSED** [review]

**Review by drasko — CHANGES_REQUESTED** [review]
Actually, I just noticed that documentation needs to be changed to reflect new way of installing source code.

@1995parham please go through the doc and identify and change it where needed. For example [here[(https://mainflux.readthedocs.io/en/latest/dev-guide/#getting-mainflux) I think you definitely need change.

**Comment by 1995parham** [comment]
@drasko Yes, you are right. I will check it.

**Review by dborovcanin — CHANGES_REQUESTED** [review]
@1995parham Please take a look at the `Readme` files for the most of our services (e.g. `bootstrap/README.md` or `http/README.md`) and update this part:
```bash
# download the latest version of the service
go get github.com/mainflux/mainflux

cd $GOPATH/src/github.com/mainflux/mainflux

# compile the service
make <SVC>
```

**Comment by drasko** [comment]
@1995parham plase take into account @dusanb94 's remark for README and also update the branch.

**Review by dborovcanin — DISMISSED** [review]

File: scripts/ci.sh
Comment (drasko) [inline scripts/ci.sh]: Not properly indented.

**Review by drasko — CHANGES_REQUESTED** [review]

**Review by drasko — DISMISSED** [review]
LGTM

**Comment by drasko** [comment]
@1995parham please update the branch

File: http/README.md
Comment (manuio) [inline http/README.md:49]: Why that? Revert to `go get` please

**Review by manuio — CHANGES_REQUESTED** [review]

File: http/README.md
Comment (manuio) [reply http/README.md:49]: @1995parham aha, I was not sure. Thank you for the clarification

**Comment by drasko** [comment]
@1995parham we had a lot of big changes recently and now they are merged. Can you please update your branch now so that the team can merge your PR tomorrow morning.

Also, please take into account all new vendoring that might have happened in the meantime.

**Comment by 1995parham** [comment]
@drasko Yes, I will do it.

**Comment by manuio** [comment]
@1995parham be careful with DCO, it's failing.

File: scripts/ci.sh
Comment (dborovcanin) [inline scripts/ci.sh:38]: Please remove an extra space character here (`get  -u`). Also, please check why CI is failing. You can try to remove `git checkout` commands using versioning directly in  `go get`, something like:
```bash
go get  -u github.com/golang/protobuf/protoc-gen-go@v1.3.2 \
	github.com/gogo/protobuf/protoc-gen-gofast@v1.3.1 \
	google.golang.org/grpc@v1.24.0
```

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: scripts/ci.sh
Comment (1995parham) [reply scripts/ci.sh:38]: Yes, I will fix it.

Comment (drasko) [reply scripts/ci.sh:38]: @1995parham this PR is complex, touches a lot of stuf and if we want it merged must be really well maintained. You must work every day with Mainflux developers until it is merged properly, otherwise it goes out of the sync and stays hanging there for months.

I propose that either we merge it on Monday, or we close it due specific and effort-demanding task, and let one of Mainflux core members to handle this while you switch to some other PRs that would be more isolated and easier to merge.

Comment (1995parham) [reply scripts/ci.sh:38]: @drasko Thanks for your comment. I think I can switch to another Issue and let it be merged by one of the core members. I will choose another issue based on Mainflux Gitter.

**Comment by drasko** [comment]
Closing this one due to its high impact. It must be handled carefully by Mainflux core engineer.

**Pull request closed without merging**
